pkgs/coverage/lib/src/isolate_paused_listener.dart

```diff
@@ -0,0 +1,218 @@
+    await listenToIsolateLifecycleEvents(_service, _onStart, _onPause, _onExit);
+
+    await _allExitedCompleter.future;
+
+    // Resume the main isolate.
+    if (_mainIsolate != null) {
+      await _service.resume(_mainIsolate!.id!);
+    }
+  }
+
+  IsolateGroupState _getGroup(IsolateRef isolateRef) =>
+      isolateGroups[isolateRef.isolateGroupId!] ??= IsolateGroupState();
+
+  void _onStart(IsolateRef isolateRef) {
+    if (_allExitedCompleter.isCompleted) return;
+    _getGroup(isolateRef).start(isolateRef.id!);
+  }
+
+  Future<void> _onPause(IsolateRef isolateRef) async {
+    if (_allExitedCompleter.isCompleted) return;
+    final group = _getGroup(isolateRef);
+    group.pause(isolateRef.id!);
+    try {
+      await _onIsolatePaused(isolateRef, group.noRunningIsolates);
+    } finally {
+      await _maybeResumeIsolate(isolateRef);
+    }
+  }
+
+  Future<void> _maybeResumeIsolate(IsolateRef isolateRef) async {
+    if (_mainIsolate == null && _isMainIsolate(isolateRef)) {
+      _mainIsolate = isolateRef;
+      // Pretend this isolate has exited so _allExitedCompleter can complete.
+      _onExit(isolateRef);
+    } else {
+      await _service.resume(isolateRef.id!);
+    }
+  }
+
+  void _onExit(IsolateRef isolateRef) {
+    if (_allExitedCompleter.isCompleted) return;
+    _getGroup(isolateRef).exit(isolateRef.id!);
+    if (isolateGroups.values.every((group) => group.noLiveIsolates)) {
+      _allExitedCompleter.complete();
+    }
+  }
+
+  static bool _isMainIsolate(IsolateRef isolateRef) {
+    // HACK: This should pretty reliably detect the main isolate, but it's not
+    // foolproof and relies on unstable features. The Dart standalone embedder
+    // and Flutter both call the main isolate "main", and they both also list
+    // this isolate first when querying isolates from the VM service. So
+    // selecting the first isolate named "main" combines these conditions and
+    // should be reliable enough for now, while we wait for a better test.
+    // TODO(https://github.com/dart-lang/sdk/issues/56732): Switch to more
+    // reliable test when it's available.
+    return isolateRef.name == 'main';
+  }
+}
+
+/// Listens to isolate start and pause events, and backfills events for isolates
+/// that existed before listening started.
+///
+/// Ensures that:
+///  - Every [onIsolatePaused] and [onIsolateExited] call will be preceeded by
+///    an [onIsolateStarted] call for the same isolate.
+///  - Not every [onIsolateExited] call will be preceeded by a [onIsolatePaused]
+///    call, but a [onIsolatePaused] will never follow a [onIsolateExited].
+///  - [onIsolateExited] will always run after [onIsolatePaused] completes, even
+///    if an exit event arrives while [onIsolatePaused] is being awaited.
+///  - Each callback will only be called once per isolate.
+Future<void> listenToIsolateLifecycleEvents(
+    VmService service,
+    void Function(IsolateRef isolate) onIsolateStarted,
```

Comment:
Nit: consider creating a typedef for all these callback types. Maybe something like: `typedef FutureOr<void> Function(IsolateRef) IsolateLifecycleCallback`?

Reply:
Ok, I've made some typedefs, but it's important to distinguish the async callbacks from the sync ones. If all of them were allowed to be async, then ensuring they don't interleave would be more complicated.